caps/types_bool_file.go

```diff
@@ -0,0 +1,46 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package caps
+
+import (
+	"regexp"
+)
+
+// BoolFileType is a built-in capability type for files that follow a
+// simple boolean protocol. The file can be read, which yields ASCII '0'
+// (zero) or ASCII '1' (one). The same can be done for writing.
+//
+// This capability type can be used to describe many boolean flags exposed
+// in sysfs, including certain hardware like exported GPIO pins.
+var BoolFileType = &Type{
+	Name: "bool-file",
+	Attrs: map[string]TypeAttr{
+		"path": &pathAttr{
+			errorHint: "only LED brightness or GPIO value is allowed",
+			// Allowed devices include:
+			allowedPatterns: []*regexp.Regexp{
+				// The brightness of standard LED class device
+				regexp.MustCompile("^/sys/class/leds/[^/]+/brightness$"),
+				// The value of standard exported GPIO
+				regexp.MustCompile("^/sys/class/gpio/gpio[0-9]+/value$"),
```

Comment:
This seems to be adding a layer of abstraction too early. Our system has complete knowledge of all capability types supported, because there are security concerns involved, and every one of these capability types will have special needs, by definition, because they enable the system to do something new. Having a a domain-specific declarative language right now means asking capability type authors (including you and me) to fit their understanding of what a capability type should do to this new DSL which none of us are comfortable with, and that before we have any experience developing actual capabilities at all.

My recommendation would be to allow capabilities to accept, reject, and filter their respective attributes in a procedural way, and for that we only need a single Go method. We already have Validate which does some of that. Let's fix the design to allow it to do all of it. Once we have a good volume of real-world capabilities actually working using that very simple but flexible and well-understood method, I'd be happy to talk again about how to abstract away the implementation into a declarative format. We'll have a much better idea by then of what is worth doing or not.